02-household-queries/optimize_encoder.py

```diff
@@ -0,0 +1,163 @@
+import os
+from bs4 import BeautifulSoup
+import dotenv
+from langchain.docstore.document import Document
+from langchain_text_splitters import RecursiveCharacterTextSplitter #, NLTKTextSplitter, SpacyTextSplitter
+import json
+from langchain_community.embeddings import (
+    SentenceTransformerEmbeddings,
+    HuggingFaceEmbeddings,
+)
+import chromadb
+from chromadb.config import Settings
+from llm import ollama_client
+# from langchain_google_genai import ChatGoogleGenerativeAI, GoogleGenerativeAIEmbeddings
+
+from retrieval import create_retriever
+from langchain_community.vectorstores import Chroma
+# import nltk
+# import spacy
+
+dotenv.load_dotenv()
+
+_llm_model_name = os.environ.get("LLM_MODEL_NAME", "mistral")
+
+llm =  ollama_client(_llm_model_name, settings={"temperature":0.1})
+
+EMBEDDINGS = {
+     "st_all-MiniLM-L6-v2" : {"func": SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2"), "token_limit":256},
+     "hf_all-MiniLM-L6-v2" : {"func": HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2"), "token_limit":256},
+    #  "google_models/embedding-001": {"func": GoogleGenerativeAIEmbeddings(model="models/embedding-001"), "token_limit":2048},
+    #  "google_models/text-embedding-004": {"func": GoogleGenerativeAIEmbeddings(model="models/text-embedding-004"), "token_limit":768},
+     "BAAI/bge-small-en-v1.5" : {"func": SentenceTransformerEmbeddings(model_name="BAAI/bge-small-en-v1.5"), "token_limit":512},
+     "mixedbread-ai/mxbai-embed-large-v1" : {"func": SentenceTransformerEmbeddings(model_name="mixedbread-ai/mxbai-embed-large-v1"), "token_limit":1024},
+}
+
+def load_training_json():
+    with open("question_answer_citations.json", encoding="utf-8") as data_file:
+        json_data = json.load(data_file)
+        # print(json.dumps(json_data, indent=2))
+        return json_data
+
+
+def compute_percent_retrieved(retrieved_cards, guru_cards):
+    missed_cards = set(guru_cards) - set(retrieved_cards)
+    return (len(guru_cards) - len(missed_cards)) / len(guru_cards)
+
+
+def count_extra_cards(retrieved_cards, guru_cards):
+    extra_cards = set(retrieved_cards) - set(guru_cards)
+    return len(extra_cards)
+
+
+# split text into chunks
+def get_text_chunks_langchain(text, source, chunk_size, chunk_overlap, token_limit):
+    text_splitter = RecursiveCharacterTextSplitter(chunk_size=chunk_size, chunk_overlap=chunk_overlap)
+    # text_splitter= NLTKTextSplitter()
+    # text_splitter= SpacyTextSplitter()
+    texts = text_splitter.split_text(source + "\n\n" + text)
+    # print("  Split into", len(texts))
+    token_total = 0
+    for t in texts:
+        token_total+=llm.get_num_tokens(t)
+    if (token_total > token_limit):
+        print(f"{source} exceeded token size: {str(token_total)}")
```

Comment:
The `token_limit` is the limit per chunk `t`, right? If so, shouldn't the logic be: `if (llm.get_num_tokens(t) > token_limit)`?